code/_onclick/reach.dm

```diff
@@ -0,0 +1,239 @@
+/**
+ * A backwards depth-limited breadth-first-search to see if the target is
+ * logically "in" anything adjacent to us.
+ */
+/atom/movable/proc/has_direct_access_to(atom/ultimate_target, depth = INVENTORY_DEPTH)
+	if(!ismovable(ultimate_target))
+		return loc == ultimate_target
+	var/atom/movable/checking_movable = ultimate_target
+	var/list/procesing_locs = checking_movable.get_locs()
+	var/list/next_locs = list()
+	for(var/i in 1 to depth)
+		for(var/atom/checking_loc as anything in procesing_locs) // If it's a turf it has reached the end of the line.
+			if(checking_loc == src || checking_loc == loc)
+				return TRUE
+			if(!ismovable(checking_loc)) // Turfs are the end of the line.
+				continue
+			checking_movable = checking_loc
+			next_locs += checking_movable.get_locs()
+		if(!length(next_locs))
+			return FALSE
+		procesing_locs = next_locs.Copy()
+		next_locs.Cut()
```

Comment:
This would be faster as this so you're not copying a list of contents all the time
```suggestion
		procesing_locs = next_locs
		next_locs = list()
```